### DISCUSSION TITLE
What is the best way to provide alternative listeners for Axum?

### DISCUSSION THREAD

**Discussion by vi** [tokio-rs/axum#2472] [Q&A] [answered]
### Summary

Axum's most straightforward way to connect HTTP world to the socket world is [`axum::serve::serve`](https://docs.rs/axum/latest/axum/serve/fn.serve.html) function. It however uses [`tokio::net::TcpListener`](https://docs.rs/tokio/latest/tokio/net/struct.TcpListener.html) directly, without options to use e.g. [`tokio::net::UnixListener`](https://docs.rs/tokio/latest/tokio/net/struct.UnixListener.html) straightforwardly.

Even [the official UNIX socket example](https://github.com/tokio-rs/axum/blob/e63cc49f5545f0323122898c12981a7815da995b/examples/unix-domain-socket/src/main.rs) is somewhat lengthy and involved.

My library [tokio_listener](https://github.com/vi/tokio-listener) is aimed at being something like a drop-in replacement of `TcpListener`, but where you can specify e.g. UNIX path or inetd-like mode as as an alternative; and `axum::serve` being TCP-specific interferes with providing a simple API [like Axum 0.6 had](https://github.com/vi/tokio-listener/blob/main/examples/clap_axum06.rs). It is not practical to offer users users to bring in a big chunk of code similar to the UNIX socket example above.

Current (unpublished to crates.io so far) [implementation](https://github.com/vi/tokio-listener/blob/main/src/axum07.rs) just copies the code from `axum:serve` module, replacing `TcpListener` and `SocketAddr` with my types. This works, but results in a code duplication. For example, [`with_graceful_shutdown`](https://docs.rs/axum/latest/axum/serve/struct.Serve.html#method.with_graceful_shutdown) also get duplicated, though it is not directly related to socket type. This may complicate further eventual upgrades next Axum versions. **Is there maybe a better way?**

---

P.S. Why Tokio's [`unix::SocketAddr`](https://docs.rs/tokio/latest/tokio/net/unix/struct.SocketAddr.html) is not `Clone`? It results in additional ugliness to support [`Connected`](https://docs.rs/axum/latest/axum/extract/connect_info/trait.Connected.html).

### axum version

0.7.3

**Comment by davidpdrsn** [comment]
> Even [the official UNIX socket example](https://github.com/tokio-rs/axum/blob/e63cc49f5545f0323122898c12981a7815da995b/examples/unix-domain-socket/src/main.rs) is somewhat lengthy and involved.

That example is the best way even though it is a bit lengthy.

`axum::serve` intentionally only accepts `TcpListener` as that is the most common use case and [greatly simplifies the API](https://docs.rs/hyper/0.14.28/hyper/server/struct.Builder.html#method.serve).

> P.S. Why Tokio's [unix::SocketAddr](https://docs.rs/tokio/latest/tokio/net/unix/struct.SocketAddr.html) is not Clone? It results in additional ugliness to support [Connected](https://docs.rs/axum/latest/axum/extract/connect_info/trait.Connected.html).

Dunno but it appears to be a wrapper around https://docs.rs/mio/latest/mio/net/struct.SocketAddr.html which is itself a wrapper around some libc stuff that I'm not familiar with.

**Reply by vi** [reply]
> That example is the best way even though it is a bit lengthy.

Does it mean that to provide an `axum::serve`-like API one needs a copy of `axum::serve`?

**Reply by jplatte** [reply, marked as answer]
Yes. @davidpdrsn maybe we should add `serve_uds` to `axum-extra`?

**Reply by davidpdrsn** [reply]
@jplatte that's a good idea! I've made an issue for it https://github.com/tokio-rs/axum/issues/2474
